### PR TITLE
Tracing API

### PR DESCRIPTION
**Pull request by sdorra**
## Proposed changes

This PR introduces a new api to trace long running tasks such as calling external systems. 
The new api uses the wording from the open telemetry project (trace, span, etc.).
The trace api comes with an ExtensionPoint for the reporting of collected spans, called exporter.
The core comes with an exporter which logs the collected span.
The api is also integrated into the AdvancedHttpClient api.

### Your checklist for this pull request

- [X] PR is well described
- [X] Related issues linked to PR if existing and labels set
- [X] Target branch is not master (in most cases develop should bet the target of choice) 
- [X] Code does not conflict with target branch
- [X] New code is covered with unit tests
- [X] CHANGELOG.md updated
- [X] Definition of Done's fulfilled: [DoD](https://github.com/scm-manager/scm-manager/blob/develop/docs/en/development/definition-of-done.md) // [UI DoD](https://github.com/scm-manager/scm-manager/blob/develop/docs/en/development/ui-dod.md)
- [X] Documentation updated (only necessary for new features or changed behaviour)

### Checklist for branch merge request (not required for forks)

- [ ] Branch is green/blue on [Jenkins](https://oss.cloudogu.com/jenkins/)
- [ ] Quality Gate passed on [SonarQube](https://sonarcloud.io/organizations/scm-manager/projects)

### CODE REVIEW

File: scm-core/src/main/java/sonia/scm/trace/SpanContext.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+@Setter
+public class SpanContext {
+
+  String kind;
+  Map<String, String> labels;
+  @XmlJavaTypeAdapter(XmlInstantAdapter.class)
+  Instant opened;
+  @XmlJavaTypeAdapter(XmlInstantAdapter.class)
+  Instant closed;
+  boolean failed;
```
Comment (sdorra) [inline scm-core/src/main/java/sonia/scm/trace/SpanContext.java:61]: This is far too much visibility. The `SpanContext` should be immutable. I know this hard with JAXB, but it should be at least pseudo immutable (only mutable from the same package.

File: scm-webapp/src/main/java/sonia/scm/net/ahc/DefaultAdvancedHttpClient.java [hunk 2]
```diff
@@ -380,7 +405,10 @@ private boolean isProxyEnabled(BaseHttpRequest<?> request)
 
   /** set of content transformers */
   private final Set<ContentTransformer> contentTransformers;
-  
+
   /** ssl context provider */
   private final Provider<SSLContext> sslContextProvider;
+
+  /** tracer used for request tracing */
+  private final Tracer tracer;
```
Comment (eheimbuch) [inline scm-webapp/src/main/java/sonia/scm/net/ahc/DefaultAdvancedHttpClient.java:413]: I would like this fields at the top of the class

File: scm-webapp/src/test/java/sonia/scm/net/ahc/DefaultAdvancedHttpClientTest.java [hunk 3]
```diff
@@ -364,4 +417,10 @@ protected HttpURLConnection createProxyConnecton(URL url,
 
   /** Field description */
   private Set<ContentTransformer> transformers;
+
+  @Mock
+  private Tracer tracer;
+
+  @Mock
+  private Span span;
```
Comment (eheimbuch) [inline scm-webapp/src/test/java/sonia/scm/net/ahc/DefaultAdvancedHttpClientTest.java:425]: Fields at top of class

File: scm-webapp/src/test/java/sonia/scm/net/ahc/DefaultAdvancedHttpResponseTest.java [hunk 4]
```diff
@@ -130,13 +140,10 @@ public void testGetHeaders()
...
-  /** Field description */
-  private final DefaultAdvancedHttpClient client =
-    new DefaultAdvancedHttpClient(new ScmConfiguration(), new HashSet<>(), new SSLContextProvider());
-
   /** Field description */
   @Mock
   private HttpURLConnection connection;
+
+  @Mock(answer = Answers.RETURNS_DEEP_STUBS)
+  private Tracer tracer;
```
Comment (eheimbuch) [inline scm-webapp/src/test/java/sonia/scm/net/ahc/DefaultAdvancedHttpResponseTest.java:148]: Especially here you move the fields to top. One field is at top and the others at the bottom...

**Pull request merged**